File: cpp/653.cpp
class Solution {
public:
    bool findTarget(TreeNode* root, int k) {
        stack<TreeNode*> stk;
        unordered_map<int, bool> m;
        TreeNode *cur = root;
        while (cur != nullptr || !stk.empty()) {
            while (cur != nullptr) {
                stk.push(cur);
                cur = cur->left;
            }
            if (!stk.empty()) {
                cur = stk.top();
                stk.pop();
                if (m.count(k-cur->val)) 
                    return true;
                m[cur->val] = true;
                cur = cur->right;
            }
        }
        return false;
    }
};

File: go/653.go
findTarget(root *TreeNode, k int) bool {
    m, stk, cur := map[int]bool{}, []*TreeNode{}, root
    for cur != nil || len(stk) > 0 {
        for cur != nil {
            stk = append(stk, cur)
            cur = cur.Left
        }
        if len(stk) > 0 {
            cur = stk[len(stk)-1]
            if _, ok := m[k-cur.Val]; ok {
                return true
            }
            m[cur.Val] = true
            stk = stk[:len(stk)-1]
            cur = cur.Right
        }
    }
    return false
}

type TreeNode struct {
    Val   int
    Left  *TreeNode
    Right *TreeNode
}

